jwst/extract_1d/extract.py

```diff
@@ -2583,6 +2583,12 @@ def run_extract1d(input_model, refname, smoothing_length, bkg_order,
                                 log_increment, subtract_background,
                                 apply_nod_offset, was_source_model)
 
+    # Remove target.source_type from the output model, so that it
+    # doesn't force creation of an empty SCI extension in the output
+    # x1d product just to hold this keyword.
+    if hasattr(output_model.meta.target, "source_type"):
```

Comment:
Because `source_type` is in the schema, `hasattr` will always evaluate to true.

Deleting the attribute is maybe not great?  Can you set it to `None` instead?  It shouldn't get written out then.

```
In [13]: model = datamodels.ImageModel()

In [14]: model.meta.wcsinfo.roll_ref = 45.

In [15]: model.save("foo.fits")
Out[15]: 'foo.fits'

In [16]: !fitsinfo foo.fits
Filename: foo.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU       7   ()      
  1  SCI           1 ImageHDU        10   ()      
  2  ASDF          1 BinTableHDU     11   1R x 1C   [523B]   

In [17]: !fitsheader foo.fits
# HDU 0 in foo.fits:
SIMPLE  =                    T / conforms to FITS standard                      
BITPIX  =                    8 / array data type                                
NAXIS   =                    0 / number of array dimensions                     
EXTEND  =                    T                                                  
DATE    = '2020-05-02T21:16:22.311' / [yyyy-mm-ddThh:mm:ss.ss] UTC date file cre
FILENAME= 'foo.fits'           / Name of the file                               
DATAMODL= 'ImageModel'         / Type of data model                             

# HDU 1 in foo.fits:
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                    8 / array data type                                
NAXIS   =                    0 / number of array dimensions                     
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               
EXTNAME = 'SCI     '           / extension name                                 
                                                                                
        WCS parameters                                                          
                                                                                
ROLL_REF=                 45.0 / [deg] V3 roll angle at the ref point (N over E)

# HDU 2 in foo.fits:
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                  523 / length of dimension 1                          
NAXIS2  =                    1 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    1 / number of table fields                         
TTYPE1  = 'ASDF_METADATA'                                                       
TFORM1  = '523B    '                                                            
EXTNAME = 'ASDF    '           / extension name                                 

In [18]: model.meta.wcsinfo.roll_ref = None

In [19]: model.save("foo.fits")
Out[19]: 'foo.fits'

In [20]: !fitsinfo foo.fits
Filename: foo.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU       7   ()      
  1  ASDF          1 BinTableHDU     11   1R x 1C   [509B]   

In [21]: !fitsheader foo.fits
# HDU 0 in foo.fits:
SIMPLE  =                    T / conforms to FITS standard                      
BITPIX  =                    8 / array data type                                
NAXIS   =                    0 / number of array dimensions                     
EXTEND  =                    T                                                  
DATE    = '2020-05-02T21:17:21.005' / [yyyy-mm-ddThh:mm:ss.ss] UTC date file cre
FILENAME= 'foo.fits'           / Name of the file                               
DATAMODL= 'ImageModel'         / Type of data model                             

# HDU 1 in foo.fits:
XTENSION= 'BINTABLE'           / binary table extension                         
BITPIX  =                    8 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                  509 / length of dimension 1                          
NAXIS2  =                    1 / length of dimension 2                          
PCOUNT  =                    0 / number of group parameters                     
GCOUNT  =                    1 / number of groups                               
TFIELDS =                    1 / number of table fields                         
TTYPE1  = 'ASDF_METADATA'                                                       
TFORM1  = '509B    '                                                            
EXTNAME = 'ASDF    '           / extension name                                 
```

Reply:
I swear I had initially tried just setting it to None, but then got "attribute not found" errors, so I went to using "hasattr". But now that you pointed this out, I switched it all back to not using "hasattr" and just setting to None and it all seems to be working for quite a number of spectral-based regtests I've tried it on. So ... go figure. Latest commit will have the updates.

Reply:
Great!

I thought about this a bit more, and I think perhaps the more elegant solution is to not have anything in `core.schema` that gets mapped to non-PRIMARY HDUs.  Instead have a separate `keyword_foo` schema or `sci` schema that maps those, and then you can be selective of what gets written out.  The reason the x1d files are getting the keyword (possibly) in the first place is because it is in `core.schema`.  If it is not, but included in every datamodel that needs it, we can avoid having `x1d` files get it altogether.

This doesn't solve the issue that Jane was seeing in BUNIT in the ERR array of `ImageModel`, as that has to be there.  But in that case updating just the primary array via `update()` is the solution.